### PR TITLE
fix:[CDS-84663]: Storing primaryManifestId in sweeping output

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Storing primaryManifestId in sweeping output to make it accessible to the users.
Expression: <+manifestConfig.primaryManifestId>

## Tests

1. K8s deploy using this variable in shell script
Screenshot 2023-11-24 at 10 43 27 PM
Screenshot 2023-11-24 at 10 43 43 PM

2. Helm deploy using this variable in shell script
Screenshot 2023-11-24 at 10 47 18 PM
Screenshot 2023-11-24 at 10 47 30 PM

3. With concatenation of expressions
Screenshot 2023-11-28 at 8 10 27 PM
Screenshot 2023-11-28 at 8 11 59 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55945)

### CODE REVIEW

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java [hunk 1]
```diff
@@ -307,6 +308,9 @@ private Optional<ManifestsOutcome> resolveManifestsOutcome(Ambiance ambiance, NG
           ngManifestsMetadataSweepingOutput.getManifestConfigurations().getPrimaryManifestRef();
       resolve(ambiance, primaryManifestRef);
       primaryManifestId = getResolvedPrimaryManifestRef(primaryManifestRef);
+      ManifestConfigurationsOutcome outcome =
+          ManifestConfigurationsOutcome.builder().primaryManifestId(primaryManifestId).build();
+      sweepingOutputService.consume(ambiance, OutcomeExpressionConstants.MANIFESTS, outcome, StepCategory.STAGE.name());
```
Comment (abosii-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:313]: Have you tested the full flow on this? We're going later to again store io/harness/cdng/manifest/steps/ManifestsStepV2.java:217, I believe this call then will fail

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Comment by abosii-harness** [comment]
@GoelPratyush99 please add required labels to the PR and screenshots with testing scenarios

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java (hunk 1 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:313]: Not completed testing yet

Comment (GoelPratyush99) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:313]: done

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abhinavsingh1196 — APPROVED** [review]

**Review by TarunUba1 — DISMISSED** [review]

**Comment by yogesh-chauhan** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/outcome/ManifestConfigurationsOutcome.java` line 31 at r2](https://reviewable.io/reviews/harness/harness-core/55945#-NkEjB5g6a8-WvhLhMMO:-NkEjB5g6a8-WvhLhMMP:b-mnafk) ([raw file](https://github.com/harness/harness-core/blob/38d6de03f20ad50ecb6f1856b238f3ede1ccfb7c/127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/outcome/ManifestConfigurationsOutcome.java#L31)):*
> ```Java
> @TypeAlias("ManifestConfigurationsOutcome")
> @RecasterAlias("io.harness.cdng.manifestConfigs.outcome.ManifestConfigurationsOutcome")
> public class ManifestConfigurationsOutcome implements Outcome, ExecutionSweepingOutput {
> ```

does it need to implement ExecutionSweepintOutput ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java` line 313 at r1](https://reviewable.io/reviews/harness/harness-core/55945#gh-1404276303:-NkEjPMR6ohKB72nqHc-:b-wl16lm) ([raw file](https://github.com/harness/harness-core/blob/1ee4efcb534339497945a21abd9ea5188c861833/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java#L313)):*
<details><summary><i>Previously, GoelPratyush99 (Pratyush Goel) wrote…</i></summary><blockquote>

done
</blockquote></details>

Does it look like the correct place to put sweepingoutput.consume() ?
The method is "resolveManifestsOutcome". 
Maybe we can have this method return the primary manifest ID and then save the sweeping output in the `io.harness.cdng.manifest.steps.ManifestsStepV2#executeAsync` method itself 
This should improve readability

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]
> Reviewed 2 of 4 files at r1, 1 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55945)* status: 5 file(s) to review. 2 change requests. 1. unresolved discussions.

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger ut5

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55945)* status: 5 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java` line 186 at r5](https://reviewable.io/reviews/harness/harness-core/55945#-NkFy4NF0OBAj-dGSJIn:-NkFy4NF0OBAj-dGSJIo:bljpv7i) ([raw file](https://github.com/harness/harness-core/blob/b50e789a2be96813507bfb40d05bd710b111fcce/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java#L186)):*
> ```Java
>       Optional<ManifestsOutcome> manifestsOutcome = manifestSpec.get().getOptionalManifestsOutcome();
>       manifestsOutcome.ifPresent(manifests -> handleManifests(ambiance, manifests, callbackIds, logCallback));
>       String primaryManifestId = manifestSpec.get().getPrimaryManifestId();
> ```

nit: we can move `String primaryManifestId = manifestSpec.get().getPrimaryManifestId();` as well into the ` saveManifestConfigurationOutcome` method

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55945)* status: 5 file(s) to review. 2 change requests. 1. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55945)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by abosii-harness** [comment]
@GoelPratyush99 please add UTs

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java [hunk 2]
```diff
@@ -226,12 +233,16 @@ public StepResponse handleAsyncResponse(
...
       PassThroughData passThroughData) {
     final NGLogCallback logCallback = serviceEnvironmentsLogUtility.getLogCallback(ambiance, false);
-    Optional<ManifestsOutcome> manifestsOutcome = resolveManifestsOutcome(ambiance, logCallback);
+    Optional<ManifestSpec> manifestSpec = resolveManifestsOutcome(ambiance, logCallback);
 
-    manifestsOutcome.ifPresent(outcome -> saveManifestsOutcome(ambiance, outcome, new HashMap<>()));
-    manifestsOutcome.ifPresent(outcome -> saveManifestExecutionDataToStageInfo(ambiance, outcome));
+    if (manifestSpec.isPresent()) {
+      Optional<ManifestsOutcome> manifestsOutcome = manifestSpec.get().getOptionalManifestsOutcome();
+      manifestsOutcome.ifPresent(outcome -> saveManifestsOutcome(ambiance, outcome, new HashMap<>()));
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:240]: both optional.ifPresent() and optional.isPresent() are being used here

**Review by abhinavsingh1196 — APPROVED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java (hunk 2 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:240]: both are on different classes. isPresent is used in ManifestSpec class and ifPresent is used on ManifestsOutcome class

**Review by abosii-harness — APPROVED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java (hunk 2 above)
Comment (abhinavsingh1196) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/manifest/steps/ManifestsStepV2.java:240]: I meant usage should be consistent, use either not both

**Pull request merged**